azurerm/resource_arm_dns_cname_record.go

```diff
@@ -54,112 +54,112 @@ func resourceArmDnsCNameRecord() *schema.Resource {
-	}
-	if !createResponse.IsSuccessful() {
-		return fmt.Errorf("Error creating DNS CName Record: %s", createResponse.Error)
+	props := dns.RecordSetProperties{
+		Metadata:    metadata,
+		TTL:         &ttl,
+		CnameRecord: &cnameRecord,
 	}
 
-	readRequest := rivieraClient.NewRequest()
-	readRequest.Command = &dns.GetCNAMERecordSet{
-		Name:              d.Get("name").(string),
-		ResourceGroupName: d.Get("resource_group_name").(string),
-		ZoneName:          d.Get("zone_name").(string),
+	parameters := dns.RecordSet{
+		Name:                &name,
+		RecordSetProperties: &props,
 	}
 
-	readResponse, err := readRequest.Execute()
+	//last parameter is set to empty to allow updates to records after creation
+	// (per SDK, set it to '*' to prevent updates, all other values are ignored)
+	resp, err := dnsClient.CreateOrUpdate(resGroup, zoneName, name, dns.CNAME, parameters, eTag, "")
 	if err != nil {
-		return fmt.Errorf("Error reading DNS CName Record: %s", err)
+		return err
 	}
-	if !readResponse.IsSuccessful() {
-		return fmt.Errorf("Error reading DNS CName Record: %s", readResponse.Error)
+
+	if resp.ID == nil {
+		return fmt.Errorf("Cannot read DNS CNAME Record %s (resource group %s) ID", name, resGroup)
 	}
 
-	resp := readResponse.Parsed.(*dns.GetCNAMERecordSetResponse)
-	d.SetId(resp.ID)
+	d.SetId(*resp.ID)
 
 	return resourceArmDnsCNameRecordRead(d, meta)
 }
 
 func resourceArmDnsCNameRecordRead(d *schema.ResourceData, meta interface{}) error {
-	client := meta.(*ArmClient)
-	rivieraClient := client.rivieraClient
+	dnsClient := meta.(*ArmClient).dnsClient
 
 	id, err := parseAzureResourceID(d.Id())
 	if err != nil {
 		return err
 	}
 
-	readRequest := rivieraClient.NewRequestForURI(d.Id())
-	readRequest.Command = &dns.GetCNAMERecordSet{}
+	resGroup := id.ResourceGroup
+	name := id.Path["CNAME"]
+	zoneName := id.Path["dnszones"]
 
-	readResponse, err := readRequest.Execute()
+	result, err := dnsClient.Get(resGroup, zoneName, name, dns.CNAME)
 	if err != nil {
-		return fmt.Errorf("Error reading DNS A Record: %s", err)
+		return fmt.Errorf("Error reading DNS CNAME record %s: %v", name, err)
 	}
-	if !readResponse.IsSuccessful() {
-		log.Printf("[INFO] Error reading DNS A Record %q - removing from state", d.Id())
+	if result.Response.StatusCode == http.StatusNotFound {
 		d.SetId("")
-		return fmt.Errorf("Error reading DNS A Record: %s", readResponse.Error)
+		return nil
 	}
 
-	resp := readResponse.Parsed.(*dns.GetCNAMERecordSetResponse)
+	props := *result.RecordSetProperties
+	cNameRecord := *props.CnameRecord
+	cName := *cNameRecord.Cname
 
-	d.Set("name", resp.Name)
-	d.Set("resource_group_name", id.ResourceGroup)
-	d.Set("zone_name", id.Path["dnszones"])
-	d.Set("ttl", resp.TTL)
-	d.Set("record", resp.CNAMERecord.CNAME)
+	d.Set("name", name)
+	d.Set("resource_group_name", resGroup)
+	d.Set("zone_name", zoneName)
+	d.Set("ttl", *result.TTL)
+	d.Set("etag", *result.Etag)
+	d.Set("record", cName)
```

Comment:
can we check to see that `result.RecordSetProperties.CnameRecord` isn't nil here - unfortunately swagger/server-side changes could lead to Terraform crashing if we don't check - so it'd be good to change this too:

```
if props := result.RecordSetProperties; props != nil {
  if record := props.CnameRecord; record != nil {
    d.Set("record", record.Cname)
  }
}
```

(again, we shouldn't need the `*` here because `d.Set` will handle that automatically)